forc-pkg/src/source/reg/mod.rs

```diff
@@ -0,0 +1,523 @@
+pub mod index_file;
+
+use crate::{
+    manifest::{self, GenericManifestFile, PackageManifestFile},
+    source::{
+        self,
+        ipfs::{ipfs_client, Cid},
+    },
+};
+use anyhow::{anyhow, bail, Context};
+use file_location::{location_from_root, Namespace};
+use forc_tracing::println_action_green;
+use index_file::IndexFile;
+use serde::{Deserialize, Serialize};
+use std::{
+    fmt::Display,
+    fs,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
+use super::IPFSNode;
+
+const REG_DIR_NAME: &str = "registry";
+
+/// A package from the official registry.
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Deserialize, Serialize)]
+pub struct Source {
+    /// The name of the specified package.
+    pub name: String,
+    /// The base version specified for the package.
+    pub version: semver::Version,
+    /// The namespace this package resides in, if no there is no namespace in
+    /// registry setup, this will be `None`.
+    pub namespace: Namespace,
+}
+
+/// A pinned instance of the registry source.
+#[derive(Clone, Debug, Eq, Hash, PartialEq, Deserialize, Serialize)]
+pub struct Pinned {
+    /// The registry package with base version.
+    pub source: Source,
+    /// The corresponding CID for this registry entry.
+    pub cid: Cid,
+}
+
+/// A resolver for registry index hosted as a github repo.
+///
+/// Given a package name and a version, a `GithubRegistryResolver` will be able
+/// to resolve, fetch, pin a package through using the index hosted on a github
+/// repository.
+pub struct GithubRegistryResolver {
+    /// Name of the github organization holding the registry index repository.
+    repo_org: String,
+    /// Name of git repository holding the registry index.
+    repo_name: String,
+    /// The number of letters used to chunk package name.
+    ///
+    /// Example:
+    /// If set to 2, and package name is "foobar", the index file location
+    /// will be ".../fo/ob/ar/foobar".
+    chunk_size: usize,
+    /// Type of the namespacing is needed to determine whether to add domain at
+    /// the beginning of the file location.
+    namespace: Namespace,
+    /// Branch name of the registry repo, the resolver is going to be using.
+    branch_name: String,
+}
+
+/// Error returned upon failed parsing of `Pinned::from_str`.
+#[derive(Clone, Debug)]
+pub enum PinnedParseError {
+    Prefix,
+    PackageName,
+    PackageVersion,
+    Cid,
+    Namespace,
+}
+
+impl GithubRegistryResolver {
+    /// Default github organization name that holds the registry git repo.
+    const DEFAULT_GITHUB_ORG: &str = "kayagokalp";
+    /// Default name of the repository that holds the registry git repo.
+    const DEFAULT_REPO_NAME: &str = "dummy-forc.pub-index";
+    /// Default chunking size of the repository that holds registry git repo.
+    pub const DEFAULT_CHUNKING_SIZE: usize = 2;
+    /// Default branch name for the repository repo.
+    const DEFAULT_BRANCH_NAME: &str = "master";
```

Comment:
nit: branch defaulted by github is `"main"`

Reply:
Default branch name we use in fuel labs repos are master, as far as i can see. 

Also this default is for pointing to the forc.pub-index repo which uses master.Although for now it uses my dummy repo for testing